<a href="https://colab.research.google.com/github/xDanilog27/Bussiness-Inteligent/blob/main/Reto_4_%C3%81rboles_de_Decisi%C3%B3n_M%C3%A9todos_de_Clasificaci%C3%B3n_(Machine_Learning).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargar las librerias



In [ ]:
# Librerias Cientificas
import pandas as pd
import numpy as np

# Librerias Arbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. Cargar la base de datos

In [ ]:
nxl='/content/4. BD1_Longitudinal Survey_L_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)

# Seleccionar las variables
XDB=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income','HealthPlan']]

XDB = XDB.dropna()
display(XDB)

XD=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]
yd=XDB[['HealthPlan']]




Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  HealthPlan  
1          0.0         1.0  
5      40000.0         1.0  
6      25000.0         1.0  
7      27400.0         1.0  
8      52000.0         1.0  
...        ...         ...  
12557   1000.0         1.0  
12565  38000.0         1.0  
12647  43600.0         1.0  
12651  40000.0         1.0  
12678  17000.0         1.0  

[5963 rows x 9 columns]

Modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4) #mar= Modelo de Arbol -- Max_Depth=maximo 4 porque mas es muy largo
mar.fit(XD,yd) # Establecer una relacion de los datos de entrada y la salida

# Se procede con la grafica
from pydotplus import graph_from_dot_data # Crea la grafica
from sklearn.tree import export_graphviz # Exporta a PNG, JPEG y TIFF

ve=['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income'] #Nombre para el arbol - ve= Variables de entrada
dot_data=export_graphviz(mar, feature_names=ve) # Crear el grafico
graph=graph_from_dot_data(dot_data)
graph.write_png('arbol.png')

True

In [ ]:
ydp=mar.predict(XD) #Pronostio sobre la base de datos original
cm=confusion_matrix(yd,ydp) # Matriz de confusion
print("La matriz de consion es:\n",cm)

VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

Exactitutud=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud es:",Exactitutud*100)

Sensibilidad=VP/(VP+FN)
print("La sensibilidad es:",Sensibilidad*100)

Especificidad=VN/(VN+FP)
print("La especificidad es:",Especificidad*100)

Precision=VP/(VP+FP)
print("La precision es:",Precision*100)

TasaError=1-Exactitutud
print("La tasa de error es:",TasaError*100)

PrediccionNegativa=VN/(VN+FN)
print("La prediccion negativa es:",PrediccionNegativa*100)

PrediccionPositiva=VP/(VP+FP)
print("La prediccion positiva es:",PrediccionPositiva*100)


La matriz de consion es:
 [[   3  850]
 [   0 5110]]
La exactitud es: 85.74543015260775
La sensibilidad es: 100.0
La especificidad es: 0.3516998827667058
La precision es: 85.73825503355704
La tasa de error es: 14.25456984739225
La prediccion negativa es: 100.0
La prediccion positiva es: 85.73825503355704


Resultados
La matriz de confusión muestra que el modelo tiene una alta sensibilidad (100%), lo que significa que clasifica correctamente a todas las personas con plan de salud. Sin embargo, tiene una baja especificidad (0.35%), lo que indica que clasifica erróneamente a muchas personas sin plan de salud como si lo tuvieran.

Estructura del Árbol
Variable principal de decisión

La primera división del árbol se realiza con la variable Ingreso (Income ≤ 22063.0). Esto indica que el nivel de ingresos es el factor más determinante para predecir si una persona tiene un plan de salud.
Si el ingreso es menor a 22063, se siguen evaluando otras características como peso, educación de la madre y autoestima.
Si el ingreso es mayor, la clasificación se vuelve más sencilla y con menor incertidumbre.
Ramas principales

Para ingresos bajos (< 22063): Se evalúan variables como peso, educación de la madre, autoestima e ingresos específicos.
Para ingresos altos (> 22063): La mayoría de las personas con ingresos altos tienden a tener un plan de salud. Sin embargo, factores como altura y educación del padre también afectan la clasificación.